mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp

```diff
@@ -0,0 +1,521 @@
+  }
+
+  static mlir::Value getCommWorld(mlir::ModuleOp &moduleOp,
+                                  const mlir::Location loc,
+                                  mlir::ConversionPatternRewriter &rewriter) {
+    auto context = rewriter.getContext();
+    // get external opaque struct pointer type
+    auto commStructT =
+        mlir::LLVM::LLVMStructType::getOpaque("ompi_communicator_t", context);
+    mlir::StringRef name = "ompi_mpi_comm_world";
+
+    // make sure global op definition exists
+    (void)getOrDefineExternalStruct(moduleOp, loc, rewriter, name, commStructT);
+
+    // get address of symbol
+    return rewriter.create<mlir::LLVM::AddressOfOp>(
+        loc, mlir::LLVM::LLVMPointerType::get(context),
+        mlir::SymbolRefAttr::get(context, name));
+  }
+
+  static intptr_t getStatusIgnore() { return 0; }
+
+  static mlir::Value getDataType(mlir::ModuleOp &moduleOp,
+                                 const mlir::Location loc,
+                                 mlir::ConversionPatternRewriter &rewriter,
+                                 mlir::Type type) {
+    mlir::StringRef mtype;
+    if (type.isF32())
+      mtype = "ompi_mpi_float";
+    else if (type.isF64())
+      mtype = "ompi_mpi_double";
+    else if (type.isInteger(64) && !type.isUnsignedInteger())
+      mtype = "ompi_mpi_int64_t";
+    else if (type.isInteger(64))
+      mtype = "ompi_mpi_uint64_t";
+    else if (type.isInteger(32) && !type.isUnsignedInteger())
+      mtype = "ompi_mpi_int32_t";
+    else if (type.isInteger(32))
+      mtype = "ompi_mpi_uint32_t";
+    else if (type.isInteger(16) && !type.isUnsignedInteger())
+      mtype = "ompi_mpi_int16_t";
+    else if (type.isInteger(16))
+      mtype = "ompi_mpi_uint16_t";
+    else if (type.isInteger(8) && !type.isUnsignedInteger())
+      mtype = "ompi_mpi_int8_t";
+    else if (type.isInteger(8))
+      mtype = "ompi_mpi_uint8_t";
+    else
+      assert(false && "unsupported type");
+
+    auto context = rewriter.getContext();
+    // get external opaque struct pointer type
+    auto commStructT = mlir::LLVM::LLVMStructType::getOpaque(
+        "ompi_predefined_datatype_t", context);
+    // make sure global op definition exists
+    (void)getOrDefineExternalStruct(moduleOp, loc, rewriter, mtype,
+                                    commStructT);
+    // get address of symbol
+    return rewriter.create<mlir::LLVM::AddressOfOp>(
+        loc, mlir::LLVM::LLVMPointerType::get(context),
+        mlir::SymbolRefAttr::get(context, mtype));
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// When lowering the mpi dialect to functions calls certain details
+// differ between various MPI implementations. This class will provide
+// these in a generic way, depending on the MPI implementation that got
+// selected by the DLTI attribute on the module.
+//===----------------------------------------------------------------------===//
```

Comment:
```suggestion
/// When lowering the mpi dialect to functions calls certain details
/// differ between various MPI implementations. This class will provide
/// these in a generic way, depending on the MPI implementation that got
/// selected by the DLTI attribute on the module.
```
This should be turned into a normal docstring for the struct.